pkg/cri/server/images/imagefs_info.go

```diff
@@ -18,34 +18,45 @@ package images
 
 import (
 	"context"
-	"time"
 
+	"github.com/containerd/containerd/pkg/cri/store/snapshot"
 	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
 )
 
 // ImageFsInfo returns information of the filesystem that is used to store images.
 // TODO(windows): Usage for windows is always 0 right now. Support this for windows.
 func (c *CRIImageService) ImageFsInfo(ctx context.Context, r *runtime.ImageFsInfoRequest) (*runtime.ImageFsInfoResponse, error) {
 	snapshots := c.snapshotStore.List()
-	timestamp := time.Now().UnixNano()
-	var usedBytes, inodesUsed uint64
+	snapshotterFSInfos := map[string]snapshot.Snapshot{}
+
 	for _, sn := range snapshots {
-		// Use the oldest timestamp as the timestamp of imagefs info.
-		if sn.Timestamp < timestamp {
-			timestamp = sn.Timestamp
+		if info, ok := snapshotterFSInfos[sn.Key.Snapshotter]; ok {
+			// Use the oldest timestamp as the timestamp of imagefs info.
+			if sn.Timestamp < info.Timestamp {
+				info.Timestamp = sn.Timestamp
+			}
+			info.Size += sn.Size
+			info.Inodes += sn.Inodes
+			snapshotterFSInfos[sn.Key.Snapshotter] = info
+		} else {
+			snapshotterFSInfos[sn.Key.Snapshotter] = snapshot.Snapshot{
+				Timestamp: sn.Timestamp,
+				Size:      sn.Size,
+				Inodes:    sn.Inodes,
+			}
 		}
-		usedBytes += sn.Size
-		inodesUsed += sn.Inodes
 	}
-	// TODO(random-liu): Handle content store
-	return &runtime.ImageFsInfoResponse{
-		ImageFilesystems: []*runtime.FilesystemUsage{
-			{
-				Timestamp:  timestamp,
-				FsId:       &runtime.FilesystemIdentifier{Mountpoint: c.imageFSPath},
-				UsedBytes:  &runtime.UInt64Value{Value: usedBytes},
-				InodesUsed: &runtime.UInt64Value{Value: inodesUsed},
-			},
-		},
-	}, nil
+
+	var imageFilesystems []*runtime.FilesystemUsage
+	for snapshotter, info := range snapshotterFSInfos {
```

Comment:
The kubelet uses the ImageFsInfo to handle the [image gc][1] and it always consume first one.
But the `snapshotterFSInfos` is map and then `range` for map is in random order. 
It might break the function. I suggest to return the main snapshot info in first item.
When kubelet supports multiple image fs, we can remove the restriction. 

[1]: <https://github.com/IBM/kube-safe-scheduler/blob/b7e6990864d7c5f567727f11b439632adfb50505/vendor/k8s.io/kubernetes/pkg/kubelet/images/image_gc_manager.go#L271>

Reply:
And the kubelet eviction manager also uses first one image disk info to detect the filesystem info.

https://github.com/kubernetes/kubernetes/blob/873eb57a4eec0d3898576c372eefdf20532ff1f9/pkg/kubelet/stats/cri_stats_provider.go#L420

All the eviction logic is based on **the first one** disk.

Reply:
Thanks, @fuweid. I have refined `ImageFsInfo()` to always return the default snapshotter as the first entry.

Reply:
@fuweid, I know this was merged already, but I'm trying to understand why we should fallback to the default snapshotter here, rather than using the one set in the specific runtime handler.

Reply:
@fidencio The kubelet only consumes first item and the runtime handler's snapshotter is experimental feature. Before kubelet supports the multiple imageFS items, I would like to keep it consistent (always return default snapshotter as first item). Random is not good here. It's hard to debug.

Reply:
Random is definitely not good, @fuweid.  But what's the limitation on providing the runtime handler's snapshotter?  Do we have access to that info from the ImageFsInfo() method?

I understand that runtime handler's snapshotter is an experimental feature, but whoever is using that feature has the expection that the snapshotter set in the runtime handler is going to be the one consumed by kubelet, and not the default one.

Reply:
Hmm.  Anyways, this may just be a red herring on what I'm debugging, as even forcing (in a very dirty and hacky way) the specific snapshotter to be returned first, I'm still facing a different behaviour than the expected one (or at least a different behaviour than when I set the snapshotter globally).

I'll keep debugging and, for now, sorry for the noise.